pennylane/transforms/core/transform_program.py

```diff
@@ -25,6 +25,70 @@
 from .transform_dispatcher import TransformContainer, TransformDispatcher, TransformError
 
 
+def _numpy_jac(*_, **__) -> qml.typing.TensorLike:
+    raise qml.QuantumFunctionError("No trainable parameters.")
+
+
+def _autograd_jac(classical_function, argnums, *args, **kwargs) -> qml.typing.TensorLike:
+    if not qml.math.get_trainable_indices(args) and argnums is None:
+        raise qml.QuantumFunctionError("No trainable parameters.")
+    return qml.jacobian(classical_function, argnum=argnums)(*args, **kwargs)
+
+
+# pylint: disable=import-outside-toplevel, unused-argument
+def _tf_jac(classical_function, argnums, *args, **kwargs) -> qml.typing.TensorLike:
+    if not qml.math.get_trainable_indices(args):
+        raise qml.QuantumFunctionError("No trainable parameters.")
+    import tensorflow as tf
+
+    with tf.GradientTape() as tape:
+        gate_params = classical_function(*args, **kwargs)
+    return tape.jacobian(gate_params, args)
+
+
+# pylint: disable=import-outside-toplevel, unused-argument
+def _torch_jac(classical_function, argnums, *args, **kwargs) -> qml.typing.TensorLike:
+    if not qml.math.get_trainable_indices(args):
+        raise qml.QuantumFunctionError("No trainable parameters.")
+    from torch.autograd.functional import jacobian
+
+    return jacobian(partial(classical_function, **kwargs), args)
+
+
+# pylint: disable=import-outside-toplevel
+def _jax_jac(classical_function, argnums, *args, **kwargs) -> qml.typing.TensorLike:
+    import jax
+
+    if argnums is None:
+        if not isinstance(args[0], jax.numpy.ndarray):
+            raise qml.QuantumFunctionError("No trainable parameters.")
+        argnums = 0
+    return jax.jacobian(classical_function, argnums=argnums)(*args, **kwargs)
+
+
+_jac_map = {
+    None: _numpy_jac,
+    "numpy": _numpy_jac,
+    "autograd": _autograd_jac,
+    "tf": _tf_jac,
+    "torch": _torch_jac,
+    "jax": _jax_jac,
+    "jax-jit": _jax_jac,
+}
+
+
+# pylint: disable=unused-argument
+def _classical_preprocessing(qnode, program, *args, argnums=None, **kwargs):
+    """Returns the trainable gate parameters for a given QNode input."""
```

Comment:
Don't we need to remove the `shots` kwarg any longer? I guess I'm not sure why we had to anyways, but this change seems unrelated to tidying up, so I'm curious :)

Reply:
`construct_tape` can accept the shots in `kwargs` and will properly put it onto the tape.  Then the program can properly take into consideration the shots.

Reply:
Nice, so is this an improvement, or maybe even a bug fix? Sorry about nagging, but do we want to test this or anything? :sweat_smile:

Reply:
Potentially an improvement. We would see a difference if the transform had different classical preprocessing based on whether or not there's shots, but I'm not sure we have any of those.